File: tests/python/test_buff_averages.py
"""
Python binding tests for Buff Averages indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path


sys.path.insert(0, str(Path(__file__).parent.parent.parent / 'target/wheels'))

try:
    import my_project as ta_indicators
except ImportError:
    pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close


class TestBuffAverages:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_buff_averages_accuracy(self, test_data):
        """Test Buff Averages matches expected values from Rust tests - mirrors check_buff_averages_accuracy"""
        close = test_data['close']
        volume = test_data['volume']


        fast_buff, slow_buff = ta_indicators.buff_averages(
            close,
            volume,
            fast_period=5,
            slow_period=20
        )

        assert len(fast_buff) == len(close)
        assert len(slow_buff) == len(close)


        expected_fast = [
            58740.30855637,
            59132.28418702,
            59309.76658172,
            59266.10492431,
            59194.11908892,
        ]

        expected_slow = [
            59209.26229392,
            59201.87047432,
            59217.15739355,
            59195.74527194,
            59196.26139533,
        ]


        assert_close(
            fast_buff[-6:-1],
            expected_fast,
            rtol=1e-3,
            msg="Buff Averages fast buffer last 5 values mismatch"
        )

        assert_close(
            slow_buff[-6:-1],
            expected_slow,
            rtol=1e-3,
            msg="Buff Averages slow buffer last 5 values mismatch"
        )

    def test_buff_averages_partial_params(self, test_data):
        """Test Buff Averages with partial parameters - mirrors check_buff_averages_partial_params"""
        close = test_data['close']
        volume = test_data['volume']


        fast_buff, slow_buff = ta_indicators.buff_averages(
            close,
            volume

        )

        assert len(fast_buff) == len(close)
        assert len(slow_buff) == len(close)

    def test_buff_averages_nan_handling(self, test_data):
        """Test Buff Averages handles NaN values correctly - mirrors check_buff_nan_prefix"""
        close = test_data['close']
        volume = test_data['volume']

        fast_buff, slow_buff = ta_indicators.buff_averages(
            close,
            volume,
            fast_period=5,
            slow_period=20
        )


        first_valid = next((i for i, x in enumerate(close) if not np.isnan(x)), 0)
        warmup_period = first_valid + 20 - 1


        assert np.all(np.isnan(fast_buff[:warmup_period])), "Expected NaN in fast buffer warmup period"
        assert np.all(np.isnan(slow_buff[:warmup_period])), "Expected NaN in slow buffer warmup period"


        assert np.all(np.isfinite(fast_buff[warmup_period:])), "Found unexpected NaN in fast buffer after warmup"
        assert np.all(np.isfinite(slow_buff[warmup_period:])), "Found unexpected NaN in slow buffer after warmup"

    def test_buff_averages_empty_input(self):
        """Test Buff Averages fails with empty input - mirrors check_buff_averages_empty_input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.buff_averages(empty, empty)

    def test_buff_averages_all_nan(self):
        """Test Buff Averages fails with all NaN values - mirrors check_buff_averages_all_nan"""
        nan_data = np.full(100, np.nan)
        nan_volume = np.full(100, np.nan)

        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.buff_averages(nan_data, nan_volume)

    def test_buff_averages_zero_period(self):
        """Test Buff Averages fails with zero period - mirrors check_buff_averages_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        volume_data = np.array([100.0, 200.0, 300.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.buff_averages(input_data, volume_data, fast_period=0)

    def test_buff_averages_period_exceeds_length(self):
        """Test Buff Averages fails when period exceeds data length - mirrors check_buff_averages_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        volume_small = np.array([100.0, 200.0, 300.0])

        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.buff_averages(data_small, volume_small, slow_period=10)

    def test_buff_averages_very_small_dataset(self):
        """Test Buff Averages fails with insufficient data - mirrors check_buff_averages_very_small_dataset"""
        single_point = np.array([42.0])
        single_volume = np.array([100.0])

        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.buff_averages(single_point, single_volume)

    def test_buff_averages_mismatched_lengths(self):
        """Test Buff Averages fails with mismatched data lengths - mirrors check_buff_averages_mismatched_lengths"""
        price_data = np.array([10.0, 20.0, 30.0])
        volume_data = np.array([100.0, 200.0])

        with pytest.raises(ValueError, match="mismatched|different length"):
            ta_indicators.buff_averages(price_data, volume_data)

    def test_buff_averages_streaming(self, test_data):
        """Test BuffAveragesStream for real-time updates - improved version"""
        stream = ta_indicators.BuffAveragesStream(fast_period=5, slow_period=20)


        prices = test_data['close'][:50]
        volumes = test_data['volume'][:50]

        stream_results = []
        for i in range(len(prices)):
            result = stream.update(prices[i], volumes[i])
            if i < 19:
                assert result is None, f"Should not have output before period at index {i}"
            else:
                assert result is not None, f"Should have output after period at index {i}"
                assert len(result) == 2, "Should return (fast_buff, slow_buff) tuple"
                stream_results.append(result)


        fast_buff_batch, slow_buff_batch = ta_indicators.buff_averages(
            prices,
            volumes,
            fast_period=5,
            slow_period=20
        )


        for i, (stream_fast, stream_slow) in enumerate(stream_results[-5:]):
            batch_idx = 45 + i
            assert_close(
                stream_fast,
                fast_buff_batch[batch_idx],
                rtol=1e-9,
                msg=f"Stream vs batch fast mismatch at index {batch_idx}"
            )
            assert_close(
                stream_slow,
                slow_buff_batch[batch_idx],
                rtol=1e-9,
                msg=f"Stream vs batch slow mismatch at index {batch_idx}"
            )

    def test_buff_averages_batch_single(self, test_data):
        """Test Buff Averages batch with single parameter set - mirrors check_buff_averages_batch_single"""
        close = test_data['close']
        volume = test_data['volume']


        result = ta_indicators.buff_averages_batch(
            close,
            volume,
            fast_range=(5, 5, 1),
            slow_range=(20, 20, 1)
        )

        assert 'fast' in result
        assert 'slow' in result
        assert 'fast_periods' in result
        assert 'slow_periods' in result


        assert result['fast'].shape[0] == 1
        assert result['slow'].shape[0] == 1
        assert result['fast'].shape[1] == len(close)
        assert result['slow'].shape[1] == len(close)


        fast_row = result['fast'][0]
        slow_row = result['slow'][0]


        fast_regular, slow_regular = ta_indicators.buff_averages(close, volume, 5, 20)

        assert_close(
            fast_row,
            fast_regular,
            rtol=1e-9,
            msg="Batch fast vs regular mismatch"
        )

        assert_close(
            slow_row,
            slow_regular,
            rtol=1e-9,
            msg="Batch slow vs regular mismatch"
        )

    def test_buff_averages_batch_multiple(self, test_data):
        """Test Buff Averages batch with multiple parameter combinations"""
        close = test_data['close'][:100]
        volume = test_data['volume'][:100]


        result = ta_indicators.buff_averages_batch(
            close,
            volume,
            fast_range=(3, 5, 2),
            slow_range=(15, 20, 5)
        )


        expected_combos = [
            (3, 15), (3, 20),
            (5, 15), (5, 20)
        ]

        assert result['fast'].shape[0] == 4
        assert result['slow'].shape[0] == 4
        assert len(result['fast_periods']) == 4
        assert len(result['slow_periods']) == 4


        for i, (fast_p, slow_p) in enumerate(expected_combos):
            assert result['fast_periods'][i] == fast_p
            assert result['slow_periods'][i] == slow_p


            fast_single, slow_single = ta_indicators.buff_averages(
                close, volume, fast_p, slow_p
            )

            assert_close(
                result['fast'][i],
                fast_single,
                rtol=1e-9,
                msg=f"Batch fast mismatch for combo ({fast_p}, {slow_p})"
            )

            assert_close(
                result['slow'][i],
                slow_single,
                rtol=1e-9,
                msg=f"Batch slow mismatch for combo ({fast_p}, {slow_p})"
            )

    def test_buff_averages_batch_edge_cases(self):
        """Test Buff Averages batch edge cases"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0,
                        11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0])
        volume = np.array([1.0] * 20)


        result = ta_indicators.buff_averages_batch(
            data,
            volume,
            fast_range=(5, 7, 10),
            slow_range=(10, 10, 1)
        )

        assert result['fast'].shape[0] == 1
        assert result['fast_periods'][0] == 5
        assert result['slow_periods'][0] == 10


        result = ta_indicators.buff_averages_batch(
            data,
            volume,
            fast_range=(5, 5, 0),
            slow_range=(10, 10, 0)
        )

        assert result['fast'].shape[0] == 1


        empty = np.array([])
        with pytest.raises(ValueError):
            ta_indicators.buff_averages_batch(
                empty,
                empty,
                fast_range=(5, 5, 1),
                slow_range=(20, 20, 1)
            )

    def test_buff_averages_batch_warmup_periods(self, test_data):
        """Test that batch processing correctly handles different warmup periods"""
        close = test_data['close'][:50]
        volume = test_data['volume'][:50]

        result = ta_indicators.buff_averages_batch(
            close,
            volume,
            fast_range=(3, 5, 2),
            slow_range=(10, 20, 10)
        )


        combos = [(3, 10), (3, 20), (5, 10), (5, 20)]

        for i, (fast_p, slow_p) in enumerate(combos):
            fast_row = result['fast'][i]
            slow_row = result['slow'][i]


            warmup = slow_p - 1


            assert np.all(np.isnan(fast_row[:warmup])), \
                f"Expected NaN in fast warmup for combo ({fast_p}, {slow_p})"
            assert np.all(np.isnan(slow_row[:warmup])), \
                f"Expected NaN in slow warmup for combo ({fast_p}, {slow_p})"


            assert np.all(np.isfinite(fast_row[warmup:])), \
                f"Unexpected NaN after fast warmup for combo ({fast_p}, {slow_p})"
            assert np.all(np.isfinite(slow_row[warmup:])), \
                f"Unexpected NaN after slow warmup for combo ({fast_p}, {slow_p})"


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: tests/wasm/test_buff_averages.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import fs from 'fs';
import { createRequire } from 'module';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const wasmBinPath = path.join(__dirname, '../../pkg/vector_ta_bg.wasm');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        const imported = await import(importPath);

        const wasmModule = (imported.default && typeof imported.default === 'object') ? imported.default : imported;
        if (typeof imported.default === 'function') {
            await imported.default();
        } else if (typeof wasmModule.default === 'function') {
            await wasmModule.default();
        } else if (typeof wasmModule.initSync === 'function' && fs.existsSync(wasmBinPath)) {
            wasmModule.initSync(fs.readFileSync(wasmBinPath));
        }
        wasm = wasmModule;
    } catch (error) {
        try {
            const require = createRequire(import.meta.url);
            wasm = require(path.join(__dirname, '../../pkg/vector_ta.js'));
        } catch {
            console.error('Failed to load WASM module. Run "wasm-pack build --target nodejs --out-name vector_ta --features wasm" first');
            throw error;
        }
    }

    testData = loadTestData();
});

test('Buff Averages accuracy', () => {

    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);


    const result = wasm.buff_averages_js(close, volume, 5, 20);


    const halfLen = close.length;
    const fast_buff = result.slice(0, halfLen);
    const slow_buff = result.slice(halfLen);

    assert.strictEqual(fast_buff.length, close.length);
    assert.strictEqual(slow_buff.length, close.length);


    const expected_fast = [
        58740.30855637,
        59132.28418702,
        59309.76658172,
        59266.10492431,
        59194.11908892,
    ];

    const expected_slow = [
        59209.26229392,
        59201.87047432,
        59217.15739355,
        59195.74527194,
        59196.26139533,
    ];


    const last6_fast = fast_buff.slice(-6);
    const last6_slow = slow_buff.slice(-6);
    const last5_fast = last6_fast.slice(0, 5);
    const last5_slow = last6_slow.slice(0, 5);

    assertArrayClose(
        last5_fast,
        expected_fast,
        1e-3,
        "Buff Averages fast buffer last 5 values mismatch"
    );

    assertArrayClose(
        last5_slow,
        expected_slow,
        1e-3,
        "Buff Averages slow buffer last 5 values mismatch"
    );
});

test('Buff Averages partial params', () => {

    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);


    const result = wasm.buff_averages_js(close, volume, 5, 20);

    assert.strictEqual(result.length, close.length * 2);
});

test('Buff Averages NaN handling', () => {

    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);

    const result = wasm.buff_averages_js(close, volume, 5, 20);

    const halfLen = close.length;
    const fast_buff = result.slice(0, halfLen);
    const slow_buff = result.slice(halfLen);


    let firstValid = 0;
    for (let i = 0; i < close.length; i++) {
        if (!isNaN(close[i])) {
            firstValid = i;
            break;
        }
    }

    const warmupPeriod = firstValid + 20 - 1;


    assertAllNaN(fast_buff.slice(0, warmupPeriod), "Expected NaN in fast buffer warmup period");
    assertAllNaN(slow_buff.slice(0, warmupPeriod), "Expected NaN in slow buffer warmup period");


    assertNoNaN(fast_buff.slice(warmupPeriod), "Found unexpected NaN in fast buffer after warmup");
    assertNoNaN(slow_buff.slice(warmupPeriod), "Found unexpected NaN in slow buffer after warmup");
});

test('Buff Averages empty input', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.buff_averages_js(empty, empty, 5, 20);
    }, /empty/i, "Should throw error for empty input");
});

test('Buff Averages all NaN', () => {

    const nanData = new Float64Array(100);
    const nanVolume = new Float64Array(100);
    nanData.fill(NaN);
    nanVolume.fill(NaN);

    assert.throws(() => {
        wasm.buff_averages_js(nanData, nanVolume, 5, 20);
    }, /All values are NaN/i, "Should throw error for all NaN values");
});

test('Buff Averages zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    const volumeData = new Float64Array([100.0, 200.0, 300.0]);

    assert.throws(() => {
        wasm.buff_averages_js(inputData, volumeData, 0, 20);
    }, /Invalid period/, "Should throw error for zero fast period");
});

test('Buff Averages period exceeds length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    const volumeSmall = new Float64Array([100.0, 200.0, 300.0]);

    assert.throws(() => {
        wasm.buff_averages_js(dataSmall, volumeSmall, 5, 10);
    }, /Invalid period|Not enough valid data/, "Should throw error for period exceeding data length");
});

test('Buff Averages very small dataset', () => {

    const singlePoint = new Float64Array([42.0]);
    const singleVolume = new Float64Array([100.0]);

    assert.throws(() => {
        wasm.buff_averages_js(singlePoint, singleVolume, 5, 20);
    }, /Invalid period|Not enough valid data/, "Should throw error for insufficient data");
});

test('Buff Averages mismatched lengths', () => {

    const priceData = new Float64Array([10.0, 20.0, 30.0]);
    const volumeData = new Float64Array([100.0, 200.0]);

    assert.throws(() => {
        wasm.buff_averages_js(priceData, volumeData, 2, 2);
    }, /mismatched|different length/i, "Should throw error for mismatched data lengths");
});

test('Buff Averages unified API', () => {

    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);

    const result = wasm.buff_averages(close, volume, 5, 20);


    assert(result.values, 'Should have values array');
    assert(typeof result.rows === 'number', 'Should have rows count');
    assert(typeof result.cols === 'number', 'Should have cols count');


    assert.strictEqual(result.rows, 2);
    assert.strictEqual(result.cols, close.length);
    assert.strictEqual(result.values.length, 2 * close.length);


    const fast_buff = result.values.slice(0, result.cols);
    const slow_buff = result.values.slice(result.cols);


    const expected_fast = [
        58740.30855637,
        59132.28418702,
        59309.76658172,
        59266.10492431,
        59194.11908892,
    ];

    const expected_slow = [
        59209.26229392,
        59201.87047432,
        59217.15739355,
        59195.74527194,
        59196.26139533,
    ];

    assertArrayClose(
        fast_buff.slice(-6, -1),
        expected_fast,
        1e-3,
        "Unified API fast buffer mismatch"
    );

    assertArrayClose(
        slow_buff.slice(-6, -1),
        expected_slow,
        1e-3,
        "Unified API slow buffer mismatch"
    );
});

test('Buff Averages batch single', () => {

    const close = new Float64Array(testData.close.slice(0, 100));
    const volume = new Float64Array(testData.volume.slice(0, 100));

    const result = wasm.buff_averages_batch(
        close,
        volume,
        [5, 5, 1],
        [20, 20, 1]
    );


    assert(result.values, 'Should have values array');
    assert(result.rows, 'Should have rows count');
    assert(result.cols, 'Should have cols count');
    assert(result.fast_periods, 'Should have fast_periods array');
    assert(result.slow_periods, 'Should have slow_periods array');


    assert.strictEqual(result.fast_periods.length, 1);
    assert.strictEqual(result.slow_periods.length, 1);
    assert.strictEqual(result.rows, 2);
    assert.strictEqual(result.cols, 100);


    assert.strictEqual(result.fast_periods[0], 5);
    assert.strictEqual(result.slow_periods[0], 20);


    const regularResult = wasm.buff_averages_js(close, volume, 5, 20);
    const fastRegular = regularResult.slice(0, 100);
    const slowRegular = regularResult.slice(100);

    const fastBatch = result.values.slice(0, 100);
    const slowBatch = result.values.slice(100);

    assertArrayClose(fastBatch, fastRegular, 1e-10, "Batch vs regular fast mismatch");
    assertArrayClose(slowBatch, slowRegular, 1e-10, "Batch vs regular slow mismatch");
});

test('Buff Averages batch multiple', () => {

    const close = new Float64Array(testData.close.slice(0, 50));
    const volume = new Float64Array(testData.volume.slice(0, 50));

    const result = wasm.buff_averages_batch(
        close,
        volume,
        [3, 5, 2],
        [15, 20, 5]
    );


    assert.strictEqual(result.fast_periods.length, 4);
    assert.strictEqual(result.slow_periods.length, 4);
    assert.strictEqual(result.rows, 8);
    assert.strictEqual(result.cols, 50);


    const expectedCombos = [
        {fast: 3, slow: 15},
        {fast: 3, slow: 20},
        {fast: 5, slow: 15},
        {fast: 5, slow: 20}
    ];

    for (let i = 0; i < expectedCombos.length; i++) {
        assert.strictEqual(result.fast_periods[i], expectedCombos[i].fast);
        assert.strictEqual(result.slow_periods[i], expectedCombos[i].slow);


        const fastRow = result.values.slice(i * 50, (i + 1) * 50);
        const slowRow = result.values.slice((4 + i) * 50, (4 + i + 1) * 50);


        const regularResult = wasm.buff_averages_js(
            close,
            volume,
            expectedCombos[i].fast,
            expectedCombos[i].slow
        );
        const fastRegular = regularResult.slice(0, 50);
        const slowRegular = regularResult.slice(50);

        assertArrayClose(
            fastRow,
            fastRegular,
            1e-10,
            `Fast mismatch for combo (${expectedCombos[i].fast}, ${expectedCombos[i].slow})`
        );
        assertArrayClose(
            slowRow,
            slowRegular,
            1e-10,
            `Slow mismatch for combo (${expectedCombos[i].fast}, ${expectedCombos[i].slow})`
        );
    }
});

test('Buff Averages batch edge cases', () => {

    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]);
    const volume = new Float64Array(20);
    volume.fill(1.0);


    const result1 = wasm.buff_averages_batch(
        data,
        volume,
        [5, 7, 10],
        [10, 10, 1]
    );

    assert.strictEqual(result1.fast_periods.length, 1);
    assert.strictEqual(result1.fast_periods[0], 5);
    assert.strictEqual(result1.slow_periods[0], 10);


    assert.throws(() => {
        wasm.buff_averages_batch(
            new Float64Array([]),
            new Float64Array([]),
            [5, 5, 1],
            [20, 20, 1]
        );
    }, /empty/i);


    assert.throws(() => {
        wasm.buff_averages_batch(
            data,
            volume,
            [5, 5],
            [20, 20, 1]
        );
    }, /must each have 3 elements/);
});

test('Buff Averages batch warmup periods', () => {

    const close = new Float64Array(testData.close.slice(0, 50));
    const volume = new Float64Array(testData.volume.slice(0, 50));

    const result = wasm.buff_averages_batch(
        close,
        volume,
        [3, 5, 2],
        [10, 20, 10]
    );


    const combos = [
        {fast: 3, slow: 10, warmup: 9},
        {fast: 3, slow: 20, warmup: 19},
        {fast: 5, slow: 10, warmup: 9},
        {fast: 5, slow: 20, warmup: 19}
    ];

    for (let i = 0; i < combos.length; i++) {
        const fastRow = result.values.slice(i * 50, (i + 1) * 50);
        const slowRow = result.values.slice((4 + i) * 50, (4 + i + 1) * 50);
        const warmup = combos[i].warmup;


        for (let j = 0; j < warmup; j++) {
            assert(isNaN(fastRow[j]),
                `Expected NaN in fast warmup at index ${j} for combo (${combos[i].fast}, ${combos[i].slow})`);
            assert(isNaN(slowRow[j]),
                `Expected NaN in slow warmup at index ${j} for combo (${combos[i].fast}, ${combos[i].slow})`);
        }


        for (let j = warmup; j < 50; j++) {
            assert(!isNaN(fastRow[j]),
                `Unexpected NaN after fast warmup at index ${j} for combo (${combos[i].fast}, ${combos[i].slow})`);
            assert(!isNaN(slowRow[j]),
                `Unexpected NaN after slow warmup at index ${j} for combo (${combos[i].fast}, ${combos[i].slow})`);
        }
    }
});

test('Buff Averages zero-copy API', () => {

    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]);
    const volume = new Float64Array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]);
    const len = data.length;


    const outPtr = wasm.buff_averages_alloc(len);
    assert(outPtr !== 0, 'Failed to allocate memory');
    wasm.buff_averages_free(outPtr, len);


    const result = wasm.buff_averages_js(data, volume, 5, 10);
    assert.strictEqual(result.length, len * 2);

    const fastResult = result.slice(0, len);
    const slowResult = result.slice(len);


    for (let i = 0; i < 9; i++) {
        assert(isNaN(fastResult[i]), `Expected NaN in fast at index ${i}`);
        assert(isNaN(slowResult[i]), `Expected NaN in slow at index ${i}`);
    }


    
});

test('Buff Averages zero-copy error handling', () => {

    const ptr = wasm.buff_averages_alloc(10);
    assert(ptr !== 0, 'Should allocate memory');
    wasm.buff_averages_free(ptr, 10);


    


    assert.throws(() => {
        wasm.buff_averages_js(
            new Float64Array([]),
            new Float64Array([]),
            5, 20
        );
    }, /empty/i, 'Should fail with empty input');

    assert.throws(() => {
        wasm.buff_averages_js(
            new Float64Array([1, 2, 3]),
            new Float64Array([1, 2]),
            5, 20
        );
    }, /mismatch|different/, 'Should fail with mismatched lengths');

    assert.throws(() => {
        wasm.buff_averages_js(
            new Float64Array([1, 2, 3]),
            new Float64Array([1, 2, 3]),
            0, 20
        );
    }, /period/, 'Should fail with zero period');
});

test('Buff Averages zero-copy large dataset', () => {

    const size = 10000;
    const data = new Float64Array(size);
    const volume = new Float64Array(size);


    for (let i = 0; i < size; i++) {
        data[i] = Math.sin(i * 0.01) * 100 + 1000;
        volume[i] = Math.random() * 1000 + 100;
    }


    const outPtr = wasm.buff_averages_alloc(size);
    assert(outPtr !== 0, 'Failed to allocate large buffer');
    wasm.buff_averages_free(outPtr, size);


    const startTime = performance.now();
    const result = wasm.buff_averages_js(data, volume, 50, 200);
    const endTime = performance.now();
    console.log(`Large dataset (${size} points): ${(endTime - startTime).toFixed(2)}ms`);

    assert.strictEqual(result.length, size * 2);

    const fastResult = result.slice(0, size);
    const slowResult = result.slice(size);


    for (let i = 0; i < 199; i++) {
        assert(isNaN(fastResult[i]), `Expected NaN at fast warmup index ${i}`);
        assert(isNaN(slowResult[i]), `Expected NaN at slow warmup index ${i}`);
    }


    for (let i = 199; i < Math.min(300, size); i++) {
        assert(!isNaN(fastResult[i]), `Unexpected NaN at fast index ${i}`);
        assert(!isNaN(slowResult[i]), `Unexpected NaN at slow index ${i}`);
    }

    
});




test('Buff Averages Context API (deprecated)', () => {

    if (!wasm.BuffAveragesContext) {
        console.log('BuffAveragesContext not available, skipping test');
        return;
    }

    const ctx = new wasm.BuffAveragesContext(5, 20);
    assert(ctx, 'Failed to create BuffAveragesContext');


    assert.strictEqual(ctx.get_warmup_period(), 19, 'Warmup period should be slow_period - 1');


    const size = 50;
    const data = new Float64Array(size);
    const volume = new Float64Array(size);

    for (let i = 0; i < size; i++) {
        data[i] = i + 1;
        volume[i] = 1;
    }


    const result = ctx.compute(data, volume);
    assert(result, 'Should compute with context');
    assert.strictEqual(result.length, size * 2);

    const fastResult = result.slice(0, size);
    const slowResult = result.slice(size);


    for (let i = 0; i < 19; i++) {
        assert(isNaN(fastResult[i]), `Expected NaN in fast at index ${i}`);
        assert(isNaN(slowResult[i]), `Expected NaN in slow at index ${i}`);
    }


    for (let i = 19; i < size; i++) {
        assert(!isNaN(fastResult[i]), `Unexpected NaN in fast at index ${i}`);
        assert(!isNaN(slowResult[i]), `Unexpected NaN in slow at index ${i}`);
    }
});

test.after(() => {
    console.log('Buff Averages WASM tests completed');
});
